### PR TITLE
Created posts/get

### PR DESCRIPTION
**Pull request by hewonjeong**
- update part of mock data. It had logical contradiction in their keys before
- update createdPosts/model because of redesigning of my post.

### CODE REVIEW

File: src/server/users/models.js [hunk 1]
```diff
@@ -179,46 +179,84 @@ export class CreatedPostManager {
...
+ *       itemKey : 'item-8523910540005-dd3860f5-b82e-473b-1234-ead0f190b000',
+ *       imageKeys :
+ *        [
+ *          'image-8523569761934-dd3860f5-b82e-473b-1234-ead0f190b000',
+ *          'image-8523569761934-dd3860f5-b82e-473b-1234-ead0f190b999'
+ *        ]
+ *    }
+ * ]
+**/
+  static combineKeySets(keySets) {
```
Comment (ovekyc) [inline src/server/users/models.js]: How about change name from set to list?

'set' sound like data structure set which has no order and also no map function just like line 219.
combinedKeySet sound like set or dictionay and unordered.

I think list is more clear than set in this case

File: src/server/users/models.js [hunk 2]
```diff
@@ -179,46 +179,84 @@ export class CreatedPostManager {
...
+    const combinedKeySets = {};
```
Comment (ovekyc) [inline src/server/users/models.js]: *Set

File: src/server/users/models.js [hunk 3]
```diff
@@ -179,46 +179,84 @@ export class CreatedPostManager {
...
+          itemKey: keySet.itemKey,
+          imageKeys: []
+        };
       }
-      item.entity = entity;
-      item.imageUrl = new S3Connector().getImageUrl(imageKey);
-      return cb(null, item);
+      // if this item is created by this user -> isCreatedByUser = true
+      combinedKeySets[keySet.itemKey].isCreatedByUser =
+      (keySet.entity === ENTITY.ITEM) || combinedKeySets[keySet.itemKey].isCreatedByUser;
```
Comment (ovekyc) [inline src/server/users/models.js]: 2 space indent is needed

File: src/server/users/models.js (hunk 2 above)
Comment (ovekyc) [reply src/server/users/models.js]: And you can use Map or Set built-in datasructure here.

File: src/server/users/models.js [hunk 4]
```diff
@@ -179,46 +179,84 @@ export class CreatedPostManager {
...
-            resolve();
-          });
-        }));
-      });
-      Promise.all(promises).then(() => {
-        return cb(null, posts);
-      }).catch((err) => {
-        return err;
+      const combinedKeySets = CreatedPostManager.combineKeySets(keySets);
+      return Promise.all(combinedKeySets.map((combinedKeySet)=>{
```
Comment (ovekyc) [inline src/server/users/models.js]: I dont understand this statement..
bcz i knew as promise.all works with only with array
Could you explain me?

File: src/test/node/users/controllers.js [hunk 5]
```diff
@@ -235,10 +235,7 @@ test('get created posts', t => {
   const expected = {
     status: 200,
     userKey: testUser.key,
-    // get a number of creatposts of test user
-    length: mockCreatedPosts.filter((post)=>{
-      return (post.key.indexOf(testUser.key) !== -1);
-    }).length,
+    length: 3,
```
Comment (ovekyc) [inline src/test/node/users/controllers.js]: Where this number come from?
You may use createdpost.length or something

File: src/test/node/users/models.js [hunk 6]
```diff
@@ -270,10 +270,7 @@ test('get created posts of a user(CreatedPostManager.getPosts)', t => {
   const testUser = mockUsers[0].value;
   const expected = {
     userKey: testUser.key,
-    // get a number of savedposts of test user
-    length: mockCreatedPosts.filter((post) => {
-      return (post.key.includes(testUser.key)) ? true : false;
-    }).length,
+    length: 3,
```
Comment (ovekyc) [inline src/test/node/users/models.js]: ditto

File: src/test/node/users/models.js [hunk 7]
```diff
@@ -291,8 +288,8 @@ test('get created posts of a user(CreatedPostManager.getPosts)', t => {
     t.equal(posts.length, expected.length,
     `should ba same length of posts array : ${posts.length}`);
     posts.map((post) => {
-      if (!post.imageUrl) {
-        t.fail('there is no imageUrl Field');
+      if (!post.imageUrls) {
```
Comment (ovekyc) [inline src/test/node/users/models.js]: Is it duplicated with model 260??

File: src/test/node/users/models.js [hunk 8]
```diff
@@ -307,18 +304,20 @@ test('get created posts of a user(CreatedPostManager.getPosts)', t => {
     t.end(err);
   });
 });
-test('get created post using key set', t => {
-  const testKeys = mockCreatedPosts[0].value;
+test('get created post using key set(CreatedPostManager.getPost)', t => {
+  const testObj = {
+    isCreatedByUser: false,
+    itemKey: mockItems[0].key,
+    imageKey: mockImages[0].key
```
Comment (ovekyc) [inline src/test/node/users/models.js]: It should be an array

File: src/test/node/users/models.js [hunk 9]
```diff
@@ -307,18 +304,20 @@ test('get created posts of a user(CreatedPostManager.getPosts)', t => {
...
-    }).value
+    itemValue: mockItems[0].value
   };
-  CreatedPostManager.getPost(testKeys, (err, value) => {
-    t.equal(value.key, expected.value.key, 'should be same item key');
-    t.equal(value.userKey, expected.value.userKey, 'should be same user key');
-    t.equal(value.hasOwnProperty('imageUrl'), true, 'should have imageUrl field');
+  CreatedPostManager.getPost(testObj).then((post)=>{
+    t.equal(post.key, expected.itemValue.key, 'should be same item key');
+    t.equal(post.userKey, expected.itemValue.userKey, 'should be same user key');
```
Comment (ovekyc) [inline src/test/node/users/models.js:325]: where is user key?

File: src/server/users/models.js (hunk 1 above)
Comment (hewonjeong) [reply src/server/users/models.js]: Right. I think vars and functions's names are ambiguous. I will try to fix it

File: src/server/users/models.js (hunk 4 above)
Comment (hewonjeong) [reply src/server/users/models.js]: Array.map function returns array that is storing values returned in map func's callback block.
In this case, map's callback function return Promise objects('fetchPost').

**Comment by hewonjeong** [comment]
[As a user, I'd like to see pictures of the same event together, So that i can recognize they are engaged in same event.](https://trello.com/c/jZEzCuBA/196-as-a-user-i-d-like-to-see-pictures-of-the-same-event-together-so-that-i-can-recognize-they-are-engaged-in-same-event)

File: src/test/node/users/controllers.js (hunk 5 above)
Comment (hewonjeong) [reply src/test/node/users/controllers.js]: it means test user's created posts length.
if we get this number, we have to check userKey and count unduplicated itemKey in mockCreatedPosts.
I thought it's too ambiguous work in test code so just used magic number here.
I will replace it as const var from mockdata file. :)

File: src/server/images/models.js [hunk 10]
```diff
@@ -53,4 +53,24 @@ export default class ImageManager {
...
+   *        imageKey: 'image-8523569763000-dd3860f5-b82e-473b-1234-ead0f190b000',
+   *        imageUrl: 'url-of-image-8523569763000-dd3860f5-b82e-473b-1234-ead0f190b000'
+   *      }
+   *    ]
+   */
+  static getImageObjList(imageKeys) {
+    const objs = [];
+    const s3 = new S3Connector();
+    imageKeys.map((imageKey)=>{
+      objs.push({imageKey, imageUrl: s3.getImageUrl(imageKey)});
```
Comment (ovekyc) [inline src/server/images/models.js]: return {imageKey, ...}

File: src/server/images/models.js [hunk 11]
```diff
@@ -53,4 +53,24 @@ export default class ImageManager {
...
+   *      {
+   *        imageKey: 'image-8523569763000-dd3860f5-b82e-473b-1234-ead0f190b000',
+   *        imageUrl: 'url-of-image-8523569763000-dd3860f5-b82e-473b-1234-ead0f190b000'
+   *      }
+   *    ]
+   */
+  static getImageObjList(imageKeys) {
+    const objs = [];
+    const s3 = new S3Connector();
+    imageKeys.map((imageKey)=>{
```
Comment (ovekyc) [inline src/server/images/models.js]: return imageKeys.map(...

File: src/server/images/models.js [hunk 12]
```diff
@@ -53,4 +53,24 @@ export default class ImageManager {
...
+   * @return {Array} objs example:
+   *    [
+   *      {
+   *        imageKey: 'image-8523569763000-dd3860f5-b82e-473b-1234-ead0f190b000',
+   *        imageUrl: 'url-of-image-8523569763000-dd3860f5-b82e-473b-1234-ead0f190b000'
+   *      }
+   *    ]
+   */
+  static getImageObjList(imageKeys) {
+    const objs = [];
```
Comment (ovekyc) [inline src/server/images/models.js]: remove it

File: src/server/images/models.js [hunk 13]
```diff
@@ -53,4 +53,24 @@ export default class ImageManager {
...
+    const s3 = new S3Connector();
+    imageKeys.map((imageKey)=>{
+      objs.push({imageKey, imageUrl: s3.getImageUrl(imageKey)});
+    });
+    return objs;
```
Comment (ovekyc) [inline src/server/images/models.js]: remove it

File: src/server/users/models.js [hunk 14]
```diff
@@ -179,46 +178,87 @@ export class CreatedPostManager {
...
-        return err;
+    .then((posts) => {
+      const combinedPosts = CreatedPostManager.combinePosts(posts);
+      return Promise.all(combinedPosts.map((post)=>{
+        return CreatedPostManager.fetchPost(post).then((fetchedPost)=>{
+          if (fetchedPost.state !== STATE.REMOVED) {
+            res.push(fetchedPost);
+          }
+        });
+      })).then(()=>{
```
Comment (ovekyc) [inline src/server/users/models.js]: You can chain this then statement to 259 line

File: src/test/node/users/models.js [hunk 15]
```diff
@@ -291,8 +288,11 @@ test('get created posts of a user(CreatedPostManager.getPosts)', t => {
     t.equal(posts.length, expected.length,
     `should ba same length of posts array : ${posts.length}`);
     posts.map((post) => {
-      if (!post.imageUrl) {
-        t.fail('there is no imageUrl Field');
+      if (!post.images) {
+        t.fail('there is no imageUrls Field');
```
Comment (ovekyc) [inline src/test/node/users/models.js]: imageUrls -> images

Comment (ovekyc) [reply src/test/node/users/models.js]: And t.fail does not print message.
Please use t.comment and t.fail

File: src/server/images/models.js (hunk 12 above)
Comment (hewonjeong) [reply src/server/images/models.js]: good! thx!

File: src/server/users/models.js [hunk 16]
```diff
@@ -179,46 +178,87 @@ export class CreatedPostManager {
...
-      item.entity = entity;
-      item.imageUrl = new S3Connector().getImageUrl(imageKey);
-      return cb(null, item);
+      // if this item is created by this user -> isCreatedByUser = true
+      objMap.get(post.itemKey).isCreatedByUser =
+        (post.entity === ENTITY.ITEM) || objMap.get(post.itemKey).isCreatedByUser;
+      objMap.get(post.itemKey).imageKeys.push(post.imageKey);
     });
+    // convert map to array for return
+    for (const value of objMap.values()) {
```
Comment (festinalente91) [inline src/server/users/models.js]: check this out!  you can just code like return Array.from(objMap.values());
[https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Array/from](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Array/from)

**Comment by ovekyc** [comment]
lgtm

**Comment by ovekyc** [comment]
lgtm 

PS, ref this javascript doc
https://github.com/uber/deck.gl/blob/master/src/lib/attribute-manager.js

**Review by ovekyc — APPROVED** [review]

**Pull request merged**